From 64f5b7a82e54d86c0268c190694c7ba30ab77f8c Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 9 Nov 2015 12:24:15 -0700
Subject: [PATCH] Dont create empty regions and chunks.

Plugins like Dynmap can end up creating tons of emtpy Region Files when using chunkExists()
Prevent bukkit from creating chunks just to unload it

diff --git a/src/main/java/net/minecraft/server/ChunkRegionLoader.java b/src/main/java/net/minecraft/server/ChunkRegionLoader.java
index 206e3aa..6ab287e 100644
--- a/src/main/java/net/minecraft/server/ChunkRegionLoader.java
+++ b/src/main/java/net/minecraft/server/ChunkRegionLoader.java
@@ -36,7 +36,10 @@ public class ChunkRegionLoader implements IChunkLoader, IAsyncChunkSaver {
             }
         }
 
-        return RegionFileCache.a(this.d, i, j).chunkExists(i & 31, j & 31);
+        // TacoSpigot Start - Don't create empty region files
+        final RegionFile region = RegionFileCache.a(this.d, i, j, false);
+        return region != null && region.chunkExists(i & 31, j & 31);
+        // TacoSpigot End
     }
     // CraftBukkit end
 
diff --git a/src/main/java/net/minecraft/server/RegionFileCache.java b/src/main/java/net/minecraft/server/RegionFileCache.java
index 5528019..6a56559 100644
--- a/src/main/java/net/minecraft/server/RegionFileCache.java
+++ b/src/main/java/net/minecraft/server/RegionFileCache.java
@@ -12,7 +12,13 @@ public class RegionFileCache {
 
     public static final Map<File, RegionFile> a = Maps.newHashMap(); // Spigot - private -> public
 
+    // TacoSpigot start - add a variant that donesn't create non-existent region files
     public static synchronized RegionFile a(File file, int i, int j) {
+        return a(file, i, j, true);
+    }
+
+    public static synchronized RegionFile a(File file, int i, int j, boolean create) {
+        // TacoSpigot end
         File file1 = new File(file, "region");
         File file2 = new File(file1, "r." + (i >> 5) + "." + (j >> 5) + ".mca");
         RegionFile regionfile = (RegionFile) RegionFileCache.a.get(file2);
@@ -20,6 +26,7 @@ public class RegionFileCache {
         if (regionfile != null) {
             return regionfile;
         } else {
+            if (!create && !file2.exists()) return null; // TacoSpigot
             if (!file1.exists()) {
                 file1.mkdirs();
             }
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index b154707..70adf15 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -202,7 +202,10 @@ public class CraftWorld implements World {
             return false;
         }
 
-        net.minecraft.server.Chunk chunk = world.chunkProviderServer.getOrCreateChunk(x, z);
+        // TacoSpigot Start -- don't create chunks when trying to unload
+        net.minecraft.server.Chunk chunk = world.chunkProviderServer.getChunkIfLoaded(x, z);
+        if (chunk == null) return false;
+        // TacoSpigot End
         if (chunk.mustSave) {   // If chunk had previously been queued to save, must do save to avoid loss of that data
             save = true;
         }
-- 
2.6.1.windows.1

